### ISSUE TITLE
Web Push Notifications

### ISSUE THREAD

**Issue by mshabarov** [vaadin/platform#4037]
## Description
Vaadin provides an API in Flow to send [Web Push Notifications](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) from the server to the client's browsers, even when they are not active on the Vaadin app's website.

## Use cases
As an application administrator / moderator
I want to be able to send Web Push Notifications from my Vaadin application to the users
So that they get notified about application's essential information in real time, e.g:
 - an error occurred on the production line
 - new order received
 - supervisor approved a request
 - other information potentially useful for business app users.

## Acceptance criteria
- [x] Flow provides an API for subscribing/unsubscribing users for/from getting notifications, e.g. user click on the "Subscribe" button and server subscribed him to notification service with this API.
- [x] Flow should provide an information (e.g. subscription ID) of the subscribed user and provide API for getting this info, so that developers can persist it on their backend.
- [x] Flow provides an API to send push notifications from server to browsers to a certain users by a given IDs.
- [x] Flow provides an API to set the following information to the notification: title and body message. 

## For future versions:
- [ ] Notifications might have a link to give users opportunity to navigate to a particular page/content on the web site/mobile app relevant to this notification, so when the user clicks on the notification, browser redirects to a page containing relevant content.
- [ ] Updating or removing pending notifications, so that user can see one notification with: "You have 3 unread messages". The ability to replace a notification: replace "You have 3 unread messages" with "You have 7 unread messages" instead of adding a new notification for each message.
- [ ] Dropdown component shown on the notification body with customisable list of clickable links.
- [ ] API for seeing whether sending notification is available for the user (feature detection in browsers, allowed or blocked by user)
- [ ] The ability to clear the pending notifications in browser

## General criteria

- [x] APIs reviewed 

- [x] Design 

- [ ] Performance 

- [x] UX/DX tests in Alpha 

- [x] Documentation: [online tutorial](https://vaadin.com/docs/latest/configuration/setting-up-webpush) , [Full-stack CRM based example](https://github.com/vaadin/flow-crm-tutorial/tree/feature/webpush), [base starter](https://github.com/vaadin/base-starter-flow-webpush)

- [ ] How to test? 

- [ ] Limitations: 

## Security
- [ ] Security implications have been taken into account (elaborate or link to product security requirement specification if there are implications)

**Comment by marcushellberg** [comment]
> Flow should provide an information about user's subscription in the events being sent from browser to server, so as developers can then store them in their back-end.

This is a bit unclearly stated. What's important is that you can persist the subscription information for a given user, so that you can message them later. 

> get notified about news, marketing offerings, target advertisement and other information in real time.

I don't believe these will be the main use cases for Flow users in business apps. The use cases may be more along the line of
- An error occurred on the production line
- New order received
- Supervisor approved your request
- ...

> Notifications might have a link to give users opportunity to navigate to a particular page/content on the web site/mobile app relevant to this notification.

When opening the link, be sure to look if the user already has an instance open. Ideally, the order of preference for opening the link would be:
1. In an open, installed instance of the application
2. In an open browser tab that is already active
3. Open the installed app to the right page 
4. Open a new browser tab to the right page

I'm not sure it's possible to discern between installed/browser, but we should look into it.

**Comment by tarekoraby** [comment]
> This is a bit unclearly stated. What's important is that you can persist the subscription information for a given user, so that you can message them later.

I like how things are currently stated. Clow shouldn't meddle with how the subscription info is persisted. Flow should just provide the subscription info, persisting them should be the developer's responsibility.

**Comment by marcushellberg** [comment]
I didn't mean that Flow should handle the persistence, but that the subscription is in a format that can be persisted.

**Comment by knoobie** [comment]
Should this https://github.com/vaadin/flow/issues/6293 be part of this enhancement to make sure that flow developer instantly know that the flow tab was closed and they don't have to send additional notifications?

**Comment by tarekoraby** [comment]
A key feature of Web Push notifications is that they can appear even if the UI is not opened at all. The notification display is natively handled by the browser.

Thus, https://github.com/vaadin/flow/issues/6293bis orthogonal to the Web Push feature.

**Comment by marcushellberg** [comment]
I think that https://github.com/vaadin/flow/issues/6293 is quite relevant for web push in many instances. 

Imagine a chat application. While a user is actively chatting, it does not make sense to send a push notification for each message. But when they close the window, they may want to start receiving notifications. Without built-in support for Beacon API many users of Web Push API will end up having to implement it themselves once they get past a hello-world stage.

**Comment by tarekoraby** [comment]
Good point. So https://github.com/vaadin/flow/issues/6293 is not completely orthogonal to Web Push support. Still, let's keep this AC focused on adding Web Push support only. We should consider adding support for Beacon API in a separate effort in the future.

**Comment by mshabarov** [comment]
> This is a bit unclearly stated. What's important is that you can persist the subscription information for a given user, so that you can message them later.
> ...
> I don't believe these will be the main use cases for Flow users in business apps. The use cases may be more along the line of
> 
> * An error occurred on the production line
> * New order received
> * Supervisor approved your request
> * ...

ACs clarified to take into account these two comments. Thanks!

**Comment by mshabarov** [comment]
Without a Beacon API implemented, Web Push Notification feature may have a limited usage (e.g. for the chat example above it really needs a Beacon API), but I'd consider Beacon API feature outside of the acceptance criteria described in this ticket.

**Comment by emarc** [comment]
Seeing whether notifications are possible (_feature detection_),  getting the permission state (_granted/denied etc_), and _error reporting/handling_ should probably be included (not sure if this needs to be explicitly listed here)?

**Comment by emarc** [comment]
The ability to [replace a notification](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API#replacing_existing_notifications) would be great (the ability to include "tag" in the notification, apparently). 
Use-case (if I understand the API correctly, I have not tried): replace "You have 3 unread messages" with "You have _7_ unread messages" instead of adding a new notification for each message.)

**Comment by marcushellberg** [comment]
We may also want to consider adding support for badging API while we're implementing this so developers could also show a little red "7" badge on the app icon to indicate those 7 unread messages. https://developer.mozilla.org/en-US/docs/Web/API/Badging_API
